Add lightning-on-taproot review to bitcoinplusplus/layer-2 (#400)
bitcointranscripts/bitcointranscripts: bitcoinplusplus/layer-2/lightning-on-taproot.md

And we tweak that because if we weren't to tweak that, then somebody would trivially be able to apply the modular multiplicative inverse to lowercase s here and extract Alice's private key, which, of course, we see to avoid.
But even doing everything correctly here, there are some risks with opening, with creating two of two signatures, for example.
One of those risks is naive key aggregation.
The other one that I'm gonna talk about later is not so used, but what do I mean by naive key aggregation?
So, let's say that we have Alice and Bob, who are trying to open a channel between the two of them.
And specifically, they're trying to figure out what their shared public key is gonna be.
Obviously, somebody has gotta start in that protocol.
So, let's say that Alice initiates the conversation, and she sends her public key to Bob, and then Bob is supposed to send his public key to Alice, and the sum of those two pub keys is going to be the pub key that's going to be the shared public key for the store signature.

> Well, Bob can quite trivially do an attack where having received Alice's public key first, he first generates his own pub key, but then instead of sending his random pub key back to Alice, he subtracts Alice's pub key.

Reviewer: "pubkey(s)" is the correct word. 
We should probably include some type of glossary inside the editor to assist with this kind of cases.

And then as a result, we have that the sum of the two pub keys that they calculate is one that completely eliminates Alice's.
So, It ends up that Bob has the capability to unilaterally create a signature.
And of course, you do not want lightning channels where one of the parties is able to just unilaterally close it and unilaterally update the state without the other parties buying.
So that is one of the issues that require some solution.
Another issue, of course, is nonce reuse.

> So here I have a quite simple equation where we have two separate signatures for different messages, m' and m'', but we're using the same nonce.

Reviewer: As pointed out in the Review Guidelines, you should ensure that code-related technical terms or math equations are enclosed in backticks (`) to enhance clarity. 

There are a few other places that you should do that in this transcript 
```suggestion
So here I have a quite simple equation where we have two separate signatures for different messages, `m'` and `m''`, but we're using the same nonce.
```

As you can tell, lowercase r, which is a random number, and uppercase R, which is the same random number multiplied by the generator point, are equivalent.
So, what does the attack look like?
Well first of all, we can subtract one signature from the other, and then we see that we simply — so the r's cancel one — they cancel each other out, and we end up with an equation that is just the private key multiplied with this expression that we can actually quite trivially calculate ourselves because we know the public key, we know the public nonce point, and we also know both messages.
So, knowing those, we apply the modular multiplicative inverse, and we have just solved for x, which is the private key.
So not great, and really highlights how trivial it is to attack and to extract a private key if you have two signatures that have reused the same nonce.
It's really, like I wish this equation were written out more frequently because I think people understate quite how trivial of an attack this is.
I mean you can write Python code in like two minutes.
Yes Cody?

> So, Sony got pawned, they were reusing nonces on all of their ECBSA signatures.

Reviewer: It's called ECDSA
```suggestion
So, Sony got pawned, they were reusing nonces on all of their ECDSA signatures.
```

So, In Lightning, what we do is we simply sort the pub keys lexicographically or based on their bytes.
But if you have some other sorting algorithm, or you just know the order and it's not sorted, but it's just ordered beforehand, that also works.
The important thing is just that you know which public key corresponds to which coefficient.
And then, say if you're Alice and your pub key also comes first, so your index is 1, because in math, indices don't start at 0.
If Alice's index is 1, then she simply takes her own pub key, concatenates it with a hash of the sequence of all the pub keys, which in this case would be first hers, then Bob's.
That thing is hashed, and that hash is concatenated with her own pub key, then we hash that again, and so, you know, because of elliptic curve photography, we can simply interpret hashes as big integers, you know, we take that hash, we take that hash and simply use that as a big integer that is a coefficient for her public key.

> Now, if you have been reading the Music 2 protocol, then you will know that this is not actually true, because we have an exception for the second index.

Reviewer: This refers to the [musig protocol ](https://bitcoinops.org/en/topics/musig/)

We need the same for Bob, too.
If Alice opens a channel to Bob, then of course what she needs is eventually for Bob to provide her with a partial signature.
In order for Bob to be able to do that, Alice, right out of the gate, needs to send Bob her local nonce.
And that local nonce, which is the public nonce there, is the one that Bob is going to need in his partial signature, because the commitment incorporates an aggregation of those nonces.

> So, Bob, in order to be able to calculate the music two-step, needs to know what Alice's local nonce later down the line is going to be.

Reviewer: Here "music" is also an error from AI's transcription

> So, you know, if we just follow the flow, Alice sends Bob her local nonce, as well as the remote nonce such that bob can create a partial signature for himself.

Reviewer: Why lowercase "bob" here? It's written as "Bob" in the whole transcript

It's really going to be understanding and making sure that the cryptography is sound and safe.
Now if you have a Taproot channel open and there are some other nodes in the network, then in principle What do you think?
Are those other nodes able to send a payment if they don't support Taproot through a channel somewhere in the middle of the route that is a Taproot channel, or should they not be?
Should be able to.
Should be able to.
In fact, cryptographically speaking, there isn't really anything preventing them from being able to do so, right?
Well, here's the thing about Taproot, though.
The way that gossip works today, you have signatures that match on-chain outputs, and those on-chain outputs are signed using ECDSA, with taproot channels that wouldn't work because we now have Schnorr signatures, which means that even nodes that don't support taproot, in order to so much as be aware of the fact that there are taproot channels out there available for routing, they will need to understand taproot gossip before they even support actual taproot channels themselves.
So this is one of the issues that has to be discussed.

> Now, el has a proposal up which says that for captured gossip, you have the possibility of combining and aggregating a signature across both the on-chain Bitcoin keys and the off-chain node public keys into one.

Reviewer: "el" refers to Elle Mouton

So you reduce the footprint of the gossip signature by a considerable amount.
It's just, you have to actually do that aggregation, so it will add a little extra communication that is going to be necessary between nodes just as they are opening the channel.
And you also need to make sure that nodes understand what this gossip is supposed to look like even before they necessarily support taproot.
So that is unfortunately one of the drawbacks of taproot, that by using a different signature algorithm and different on-chain footprint, unless you have support for gossip from the very beginning, you're going to have part of the network otherwise unable and even unaware of the fact that there is a part of the network that is now based on taproot.

> So just going back to the beginning, so far we have talked about how in order to leverage the benefits of Schnorr for improved privacy, to mitigate the risks of Schnorr, we have to use Music2 for key aggregation and non-key aggregation, as well as Now, because we're using Taproot and therefore we need to use Tap scripts, we have to modify where those individual spent paths are located, as well as how the gossip looks like.

Reviewer: "Tapscript" is written as one word

And that is a privacy consideration that you're facing right now already.
That is why there's talk about Gossipy 2, which would essentially be only committing to a fraction of the money that you have put up.
But that is its own can of worms.

> So honestly, I can certainly recommend watching Matt Corallo's talk from Capcom last year, where he...

Reviewer: "Capcom" is "TabConf"

Well, it was titled, Lightning Has Broken the Stock.
And One of the major breakages is privacy.
Lightning has atrocious privacy today.

> Even though with Tampered we improve on-chain privacy, if, say, Chaina lysis and therefore the IRS—not, of course, that they're bad guys, but you know what I mean.

Reviewer: ```suggestion
Even though with Taproot we improve on-chain privacy, if, say, Chainalysis and therefore the IRS—not, of course, that they're bad guys, but you know what I mean.
```

> But That being said, another aspect where we can somewhat impropriate the STTLCs, because of the issue that I mentioned earlier, where payments in hops could not be correlated.

Reviewer: We expect reviewers to catch this kind of AI's transcription errors. You are the one watching the videos so it's easier to make the connection that at this moment he is talking about "PTLCs" as shown by the speakers slides.
```suggestion
That being said, another aspect where we can somewhat improve privacy is PTLCs, because of the issue that I mentioned earlier, where payments in hops could not be correlated.
```
